proposals/20210526-model_load_test_component.md

```diff
@@ -0,0 +1,41 @@
+#### SIG TFX-Addons
+# Project Proposal
+
+**Your name:** Michal Brys
+
+**Your email:** michal.brys@openx.com
+
+**Your company/organization:** OpenX
+
+**Project name:** Model load test TFX Component
+
+## Project Description
+A TFX component performs a load test of the exported TensorFlow model because the prediction time may vary on the model type and structure.
+
+## Project Category
+Component
+
+## Project Use-Case(s)
+This component could extend a TFX InfraValidator to validate the TensorFlow model serving performance using the load test.
+The load test results will tell us:
+* if the model will meet the prediction time requirements (essential in large scale implementations),
+* if the changes in the new model will make it faster/slower comparing to the baseline model.
+
+## Project Implementation
+The load test component will integrate with TFX using the following approach:
+* The `InfraValidator TFX Pipeline Component` component creates a TensorFlow Serving endpoint.
+* The load test component spins up a pod on the Kubernetes cluster to perform the load test (send a specified volume of the prediction requests to the TensorFlow Serving endpoint).
+* The model serving performance metrics (i.e. average prediction time) and the load test specification (the traffic volume, the network protocol used, etc.) are recorded and stored as an ML Metadata artifact.
```

Comment:
How should the examples for the load test ingested? Via a reference to an example file or via ingesting the output of a 2nd ExampleGen component in the pipeline? The test set should remain constant (ideally).

Reply:
Good point, for the results consistency we may need:
- test examples (static)
- ImportExampleGen to read the test examples
- imported test examples as a component input

Reply:
FYI, ExampleGen's output is `Examples` artifact which typically contains `tf.Example` records. For serving requests we need to convert `tf.Example` to a [`PredictRequest`](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/apis/predict.proto) (or ClassificationRequest or RegressionRequest, but those are rarely used these days) and optionally convert them to json for http request.

Converting `tf.Example` to `PredictRequest` is actually very tricky part because PredictRequest is about how model consumes the example, and it can be literally anything depending on the training code. Currently InfraValidator have a limited support on this auto conversion: it only supports Predict signature that takes serialized tf.Example. In other words, model should include a layer to deserialize it and feed it to the rest of the layers.

You can follow the InfraValidator behavior, or you can directly feed PredictRequest proto or json files using Importer node with custom artifact type.

Reply:
@chongkong after doing some tests of `InfraValidator` I think we may follow this behavior.